### ISSUE TITLE
Logic error in sample code?

### ISSUE THREAD

**Issue by VRichardJP** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#853]
I am curious about the behavior of the following piece of code:
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/a22c7c336959cbc73b62a3e8041f9f8a271b0fdb/samples/Common.c#L1049-L1061

I feels to me that the line `pPendingMessageQueue = (PStackQueue) hashValue;` makes no sense inside the if since the `hashValue` is always 0 there (and `pPendingMessageQueue` is already 0 aswell). Also the value is overwritten right after by the `hashTablePut()` call.

On the other hand this line would make totally sense in a else statement, especially because the `pPendingMessageQueue` value is expected to be set after the if for the `stackQueueEnqueue()` call (in the code above it can be NULL):
```
 hashValue = 0; 
 if (STATUS_HASH_KEY_NOT_PRESENT == 
     hashTableGet(pSampleConfiguration->pPendingSignalingMessageForRemoteClient, clientIdHash, &hashValue)) { 
     CHK_STATUS(stackQueueCreate(&pPendingMessageQueue)); 
     CHK_STATUS( 
         hashTablePut(pSampleConfiguration->pPendingSignalingMessageForRemoteClient, clientIdHash, (UINT64) pPendingMessageQueue)); 
 } else {
     pPendingMessageQueue = (PStackQueue) hashValue; 
 } 
  
 pReceivedSignalingMessageCopy = MEMCALLOC(1, SIZEOF(ReceivedSignalingMessage)); 
 *pReceivedSignalingMessageCopy = *pReceivedSignalingMessage; 
```

Basically the original code will fail on the `stackQueueEnqueue()` if the same ice candidate message is received multiple times. As a consequence the memory allocated for `pReceivedSignalingMessageCopy` is never freed aswell

**Comment by MushMal** [comment]
hashValue is the retrieved value. We set it to zero as the static code analysis tools flag this as uninitialized variable by mistake.

The logic here is to try to retrieve the clientIdHash as a key. If it doesn't exist then create a stack queue to accumulate the messages till we set things up.

Resolving

**Comment by VRichardJP** [comment]
Thank you for the explanation.
It actually strengthen my feeling this part of the code is fishy:

1. this line has no effect:
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/ccb4d7f5d7c82857d91b9dcf257d449d31366318/samples/Common.c#L1081
because `pPendingMessageQueue` is NULL before this line and `hashValue` is 0 when we enter the if statement ; the hashValue is != 0 only if the client_id key is present, which is not the case when we enter the if statement

2. if the client_id key is found, `hashValue` points to that value, but `pPendingMessageQueue` value is not updated (still NULL). This means this line:
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/ccb4d7f5d7c82857d91b9dcf257d449d31366318/samples/Common.c#L1090
is the same than:
`CHK_STATUS(stackQueueEnqueue(NULL, (UINT64) pReceivedSignalingMessageCopy));`, which obviously fails and directly go to CleanUp

3. When this `stackQueueEnqueue` fails (typically the situation 2), the memory allocated to `pReceivedSignalingMessageCopy` is not freed and leaks.

I have experimented a little. I have observed that messages of type `SIGNALING_MESSAGE_TYPE_ICE_CANDIDATE` are often received multiples whenever a viewer connect to the master, which correspond to situations 2 and 3

**Comment by MushMal** [comment]
In this case, you are right, Line 1081 is not needed - it's a No-op. 

If the hashValue is found then we just allocate a message and copy the one that's called the function with and store that in the queue.

You are actually right, the memory is being accumulated but we free at the end. This is not OK. Let me fix this

**Comment by MushMal** [comment]
Actually, I lied, we do free the messages properly in submitPendingIceCandidate in LN 969. In case of an abnormal termination, the entire queue will be freed with the allocations in the free prologue.

Please let me know if you have more questions.
